File: mtae-backend/src/functions/updateProfile.ts
import { DynamoDBClient } from '@aws-sdk/client-dynamodb';
import { DynamoDBDocumentClient, UpdateCommand } from '@aws-sdk/lib-dynamodb';
import { APIGatewayProxyEvent, APIGatewayProxyResult } from 'aws-lambda';

const client = new DynamoDBClient({});
const dynamo = DynamoDBDocumentClient.from(client);
const TABLE_NAME = process.env.TABLE_NAME!;

const headers = {
    'Content-Type': 'application/json',
    'Access-Control-Allow-Origin': '*',
    'Access-Control-Allow-Headers': 'Content-Type,Authorization',
};

export const handler = async (event: APIGatewayProxyEvent): Promise<APIGatewayProxyResult> => {
    try {
        // get userid from auth context
        const userId = event.requestContext.authorizer?.claims?.sub;
        if (!userId) {
            return { statusCode: 401, headers, body: JSON.stringify({ message: 'Unauthorized' }) };
        }

        const body = JSON.parse(event.body ?? '{}');

        // update fields (currently just display name)
        const updatedDisplayName: string = body.updatedDisplayName;
        if(!updatedDisplayName || typeof updatedDisplayName !== 'string') {
            return { statusCode: 400, headers, body: JSON.stringify({ message: 'updatedDisplayName is required and must be a string' }) };
        }
        
        await dynamo.send(new UpdateCommand({
            TableName: TABLE_NAME,
            Key: { userId },
            UpdateExpression: 'SET displayName = :name, updatedAt = :now, createdAt = if_not_exists(createdAt, :now)',
            ExpressionAttributeValues: {
                ':name': updatedDisplayName,
                ':now': new Date().toISOString(),
            }
        }));

        return {
            statusCode: 200,
            headers,
            body: JSON.stringify({ message: 'Display name updated successfully' }),
        }

    } catch (err) {
        console.error(err);
        return { statusCode: 500, headers, body: JSON.stringify({ message: 'Internal server error' }) };
    }
}

File: mtae-backend/src/lambdas/update_profile.py
import json
import os
from datetime import datetime, timezone
import boto3 # type: ignore
# ^ preinstalled in AWS Lambda Python environment

dynamodb = boto3.resource('dynamodb')
table = dynamodb.Table(os.environ['TABLE_NAME'])

HEADERS = {
    'Content-Type': 'application/json',
    'Access-Control-Allow-Origin': '*',
    'Access-Control-Allow-Headers': 'Content-Type,Authorization',
}

def handler(event, context):
    try:
        user_id = (event.get('requestContext', {}).get('authorizer', {}) or {}).get('claims', {}).get('sub')
        if not user_id:
            return {
                'statusCode': 401,
                'headers': HEADERS,
                'body': json.dumps({'message': 'Unauthorized'})
            }

        body = json.loads(event.get('body') or '{}')
        updated_display_name = body.get('updatedDisplayName')

        if not updated_display_name or not isinstance(updated_display_name, str):
            return {
                'statusCode': 400,
                'headers': HEADERS,
                'body': json.dumps({'message': 'updatedDisplayName is required and must be a string'})
            }

        now = datetime.now(timezone.utc).isoformat().replace('+00:00', 'Z')

        table.update_item(
            Key={'userId': user_id},
            UpdateExpression='SET displayName = :name, updatedAt = :now, createdAt = if_not_exists(createdAt, :now)',
            ExpressionAttributeValues={
                ':name': updated_display_name,
                ':now':  now,
            }
        )

        return {
            'statusCode': 200,
            'headers': HEADERS,
            'body': json.dumps({'message': 'Display name updated successfully'})
        }

    except Exception as e:
        print(e)
        return {
            'statusCode': 500,
            'headers': HEADERS,
            'body': json.dumps({'message': 'Internal server error'})
        }
